### PR TITLE
`bond_extra` for nomination pools

### PR DESCRIPTION
**Pull request by kianenigma**
closes https://github.com/paritytech/substrate/issues/10903

- [ ] benchmarks

### CODE REVIEW

File: frame/nomination-pools/src/lib.rs [hunk 1]
```diff
@@ -333,10 +333,21 @@ pub enum ConfigOp<T: Default + Codec + Debug> {
...
+	/// Someone is adding more funds later to this pool.
+	Later,
+}
+
+/// How to increase the bond of a delegator.
+#[derive(Encode, Decode, Clone, Copy, Debug, PartialEq, Eq, TypeInfo)]
+pub enum BondExtra<Balance> {
+	/// Take from the free balance.
+	FreeBalance(Balance),
+	/// Take the entire from the accumulated rewards.
```
Comment (emostov) [inline frame/nomination-pools/src/lib.rs:349]:
```suggestion
	/// Take the entire amount from the accumulated rewards.
```

File: frame/nomination-pools/src/lib.rs [hunk 2]
```diff
@@ -990,16 +997,24 @@ pub mod pallet {
...
 
+	/// Events of this pallet.
 	#[pallet::event]
 	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
 	pub enum Event<T: Config> {
+		/// A pool has been created.
 		Created { depositor: T::AccountId, pool_id: PoolId },
-		Joined { delegator: T::AccountId, pool_id: PoolId, bonded: BalanceOf<T> },
+		/// A delegator has became bonded in a pool.
+		Bonded { delegator: T::AccountId, pool_id: PoolId, bonded: BalanceOf<T>, joined: bool },
```
Comment (emostov) [inline frame/nomination-pools/src/lib.rs:1007]: What about two separate events `Bonded` and `BondedExtra`? I think its more clear and will mean the events take 1 byte less space

Comment (kianenigma) [reply frame/nomination-pools/src/lib.rs:1007]: I prefer a single event with `joined: bool`, but happy to go either way.

File: frame/nomination-pools/src/lib.rs [hunk 3]
```diff
@@ -1103,7 +1119,55 @@ pub mod pallet {
...
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn bond_extra(origin: OriginFor<T>, extra: BondExtra<BalanceOf<T>>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let (mut delegator, mut bonded_pool, mut reward_pool) =
+				Self::get_delegator_with_pools(&who)?;
+
+			let (points_issued, bonded) = match extra {
+				BondExtra::FreeBalance(amount) => {
+					let amount = amount.min(T::Currency::free_balance(&who));
```
Comment (emostov) [inline frame/nomination-pools/src/lib.rs:1149]: I don't think this min will work. We always do transfer keep alive for delegators (including in `try_bond_funds`), so we can't transfer all of their free balance. IMO we should always just fail if they don't have the amount

Comment (kianenigma) [reply frame/nomination-pools/src/lib.rs:1149]: yeah I also don't really like it, it is magic logic hidden in the code. I was trying to mimic the staking `bond_extra`, which does this. 

If we actually want to `.min()`, we do: 

```suggestion
					let amount = amount.min(T::Currency::free_balance(&who) - ED);
```

File: frame/nomination-pools/src/lib.rs [hunk 4]
```diff
@@ -1717,21 +1800,19 @@ impl<T: Config> Pallet<T> {
 	}
 
 	/// If the delegator has some rewards, transfer a payout from the reward pool to the delegator.
-	///
-	/// # Note
-	/// This will persist updates for the reward pool to storage. But it will *not* persist updates
-	/// to the `delegator` or `bonded_pool` to storage, that is the responsibility of the caller.
+	// Emits events and potentially triggers the pool state if the math becomes corrupt, but does
```
Comment (emostov) [inline frame/nomination-pools/src/lib.rs:1803]:
```suggestion
	// Emits events and potentially modifies pool state if any arithmetic saturates, but does
```

File: frame/nomination-pools/src/lib.rs (hunk 3 above)
Comment (kianenigma) [reply frame/nomination-pools/src/lib.rs:1152]: removed in 123234233651f936af9988746e38bb3d78141daa

**Comment by kianenigma** [comment]
Should be ready from my side.

**Review by emostov — APPROVED** [review]

**Pull request merged**